crates/turbopack-node/src/source_map/mod.rs

```diff
@@ -1,7 +1,371 @@
+    } else {
+        Style::new()
+    };
+    let highlight = if ansi_colors {
+        Style::new().bold().underline()
+    } else {
+        Style::new()
+    };
+    match resolved {
+        Err(err) => {
+            // There was an error resolving the source map
+            write!(writable, "\n    at {}", original_frame)?;
+            if *first_error {
+                write!(writable, "\n    (error resolving source map: {})", err)?;
+                *first_error = false;
+            } else {
+                write!(writable, "    (error resolving source map)")?;
+            }
+        }
+        Ok(ResolvedSourceMapping::NoSourceMap) | Ok(ResolvedSourceMapping::Unmapped) => {
+            // There is no source map for this file or no mapping for the line
+            write!(
+                writable,
+                "\n    {}",
+                format_args!("[at {}]", original_frame).style(lowlight)
+            )?;
+        }
+        Ok(ResolvedSourceMapping::Mapped { frame }) => {
+            // There is a mapping to something outside of the project (e. g. plugins,
+            // internal code)
+            write!(
+                writable,
+                "\n    {}",
+                format_args!("at {} [{}]", frame, original_frame.with_name(None)).style(lowlight)
+            )?;
+        }
+        Ok(ResolvedSourceMapping::MappedProject {
+            frame,
+            project_path,
+            lines,
+        }) => {
+            // There is a mapping to a file in the project directory
+            if let Some(name) = frame.name.as_ref() {
+                write!(
+                    writable,
+                    "\n    at {name} ({}) {}",
+                    frame
+                        .with_name(None)
+                        .with_path(&project_path.path)
+                        .style(highlight),
+                    format_args!("[{}]", original_frame.with_name(None)).style(lowlight)
+                )?;
+            } else {
+                write!(
+                    writable,
+                    "\n    at {} {}",
+                    frame.with_path(&project_path.path).style(highlight),
+                    format_args!("[{}]", original_frame.with_name(None)).style(lowlight)
+                )?;
+            }
+            let (line, column) = frame.get_pos().unwrap_or((0, 0));
+            if let FileLinesContent::Lines(lines) = &*lines {
+                let lines = lines.iter().map(|l| l.content.as_str());
+                let ctx = get_source_context(lines, line - 1, column - 1, line - 1, column - 1);
+                if ansi_colors {
+                    writable.write_char('\n')?;
+                    format_source_context_lines(&ctx, writable);
+                } else {
+                    write!(writable, "\n{}", ctx)?;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn format_source_context_lines(ctx: &SourceContextLines, f: &mut impl Write) {
```

Comment:
This is similar to the code in `crates/turbopack-cli-utils/src/issue.rs`. Can we share some code between the two?

My comments in `src/issue.rs` also apply here.